packages/relayer-lib/src/tx_builder/eth_to_cosmos.rs

```diff
@@ -399,26 +344,123 @@ where
+                        .finalized_header
+                        .execution
+                        .block_number,
+                    header.consensus_update.finalized_header.beacon.slot,
+                )
             })
             .ok_or_else(|| anyhow::anyhow!("No headers found"))?;
 
+        Ok((proof_block_number, proof_slot, headers))
+    }
+}
+
+#[async_trait::async_trait]
+impl<P> TxBuilderService<EthEureka, CosmosSdk> for TxBuilder<P>
+where
+    P: Provider + Clone,
+{
+    #[tracing::instrument(skip_all)]
+    async fn relay_events(
+        &self,
+        src_events: Vec<EurekaEventWithHeight>,
+        dest_events: Vec<EurekaEventWithHeight>,
+        target_client_id: String,
+    ) -> Result<Vec<u8>> {
+        let latest_block_number = self.eth_client.get_block_number().await?;
+        let minimum_block_number = if dest_events.is_empty() {
+            let latest_block_from_events = src_events.iter().filter_map(|e| e.block_number).max();
+            latest_block_from_events.unwrap_or(latest_block_number)
+        } else {
+            // If we have destination events (e.g. timeout), use the latest block number
+            latest_block_number
+        };
+
+        let target_height = Height {
+            revision_number: 0,
+            revision_height: minimum_block_number,
+        };
+
+        let now_since_unix = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH)?;
+
+        let mut timeout_msgs = cosmos::target_events_to_timeout_msgs(
+            dest_events,
+            &target_client_id,
+            &target_height,
+            &self.signer_address,
+            now_since_unix.as_secs(),
+        );
+
+        let (mut recv_msgs, mut ack_msgs) = cosmos::src_events_to_recv_and_ack_msgs(
+            src_events,
+            &target_client_id,
+            &target_height,
+            &self.signer_address,
+            now_since_unix.as_secs(),
+        );
+
+        let ethereum_client_state = self.ethereum_client_state(target_client_id.clone()).await?;
+        let ethereum_consensus_state = self
+            .ethereum_consensus_state(target_client_id.clone(), 0)
+            .await?;
+
+        tracing::info!(
+            "Relaying events from Ethereum to Cosmos for client {}, target block number: {}, client state latest slot: {}, consensus state slot: {}, current sync committee: {:?}, next sync committee: {:?}",
+            target_client_id,
+            minimum_block_number,
+            ethereum_consensus_state.slot,
+            ethereum_consensus_state.slot,
+            ethereum_consensus_state.current_sync_committee,
+            ethereum_consensus_state.next_sync_committee,
+        );
+
+        // get update if necessary
+        let current_beacon_block = self
+            .beacon_api_client
+            .beacon_block(&format!("{:?}", ethereum_consensus_state.slot))
+            .await?;
+        let (proof_block_number, proof_slot, headers) = if minimum_block_number
```

Comment:
Honestly not super happy about the readability here.
In go, I would have just set the default first, and change them if necessary, but using mut here feels anti-rust somehow. 
wdyt?

Reply:
I think using mut is bad too. I do think you probably don't need `proof_block_number`. Wdyt?